crates/l1tx/src/filter/visitor.rs

```diff
@@ -0,0 +1,119 @@
+use bitcoin::{Block, Transaction};
+use strata_state::{
+    batch::SignedBatchCheckpoint,
+    tx::{DepositInfo, DepositRequestInfo, ProtocolOperation},
+};
+
+use super::{
+    parse_checkpoint_envelopes, parse_da_blobs, parse_deposit_requests, parse_deposits,
+    TxFilterConfig,
+};
+use crate::messages::ProtocolTxEntry;
+
+pub trait OpsVisitor {
+    // Do stuffs with `SignedBatchCheckpoint`.
+    fn visit_checkpoint(&mut self, _chkpt: SignedBatchCheckpoint) {}
+
+    // Do stuffs with `DepositInfo`.
+    fn visit_deposit(&mut self, _d: DepositInfo) {}
+
+    // Do stuffs with `DepositRequest`.
+    fn visit_deposit_request(&mut self, _d: DepositRequestInfo) {}
+
+    // Do stuffs with DA.
+    fn visit_da<'a>(&mut self, _d: impl Iterator<Item = &'a [u8]>) {}
+
+    fn collect(self) -> Vec<ProtocolOperation>;
+}
+
+pub trait BlockIndexer {
+    type Output;
+
+    fn collect(self) -> Self::Output;
+
+    fn index_tx(&mut self, txidx: u32, tx: &Transaction, config: &TxFilterConfig);
+
+    fn index_block(mut self, block: &Block, config: &TxFilterConfig) -> Self
+    where
+        Self: Sized,
+    {
+        for (i, tx) in block.txdata.iter().enumerate() {
+            self.index_tx(i as u32, tx, config);
+        }
+        self
+    }
```

Comment:
Do we have an idea of what purposes we'd have for overriding this?  It would simplify the interface to have it split out as a separate toplevel function closer to where we have the control logic for the scanning.

Reply:
I don't actually see any reason for overriding this as of now. I see what you mean here but this default implementation just looks nicely encapsulated/namespaced in the trait.